### ISSUE TITLE
escape special characters when searching on identifier field

### ISSUE THREAD

**Issue by rramosj2** [microsoft/fhir-server#1103]
**Describe the bug**
When searching for a resource based on identifier. I should be able to escape special characters
https://www.hl7.org/fhir/search.html#escaping
if I have a patient with an identifer value of XXXX|YYYYYYY|ZZZZZ
i should be able to search [base]\Patient?identifier=XXXX\|YYYYYY......
Search returns no values

**To Reproduce**
1. add an identifier to a resource(Patient/Observation) with "|" in its values
2. search for Patient?identifier=xxx\|yyyy
3. no resources are found

**Expected behavior**
the resource with the identifer value should be returned

**Actual behavior**
no resource is found

**Comment by eladiw** [comment]
@rramosj2 , @brendankowitz 
I tried reproducing this issue. 
tl;dr I believe this is not an issue since in the bug description the '|' was not escaped.

Some more details:

while running
https://localhost:44348/Patient?identifier=186cb836-2abe-4d3b-a928-325e2**,**5c8ca1c

I get the following generated query:
exec sp_executesql N'SET STATISTICS IO ON;
SET STATISTICS TIME ON;

WITH cte0 AS
(
    SELECT ResourceSurrogateId AS Sid1
    FROM dbo.TokenSearchParam
    WHERE IsHistory = 0
        AND ResourceTypeId = @p0
        AND SearchParamId = @p1
        AND (Code = @p2
        OR Code = @p3)
    
),
cte1 AS
(
    SELECT DISTINCT TOP (@p4) Sid1, 1 AS IsMatch, 0 AS IsPartial 
    FROM cte0
    ORDER BY  Sid1 ASC
)
SELECT r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(IsMatch AS bit) AS IsMatch, CAST(IsPartial AS bit) AS IsPartial, CAST(0 AS bit) AS IsRawResourceMetaSet, r.RawResource
FROM dbo.Resource r
INNER JOIN cte1
ON r.ResourceSurrogateId = cte1.Sid1
ORDER BY r.ResourceSurrogateId ASC
OPTION(RECOMPILE)',N'@p0 smallint,@p1 smallint,@p2 varchar(128),@p3 varchar(128),@p4 int',@p0=81,@p1=861,**@p2='186cb836-2abe-4d3b-a928-325e2',@p3='5c8ca1c',**@p4=11

Notice how p2 and p3 were separated since ',' was not escaped (treated as two different values).

however, when running:
https://localhost:44348/Patient?identifier=186cb836-2abe-4d3b-a928-325e2**\,**5c8ca1c

I get:
exec sp_executesql N'SET STATISTICS IO ON;
SET STATISTICS TIME ON;

WITH cte0 AS
(
    SELECT ResourceSurrogateId AS Sid1
    FROM dbo.TokenSearchParam
    WHERE IsHistory = 0
        AND ResourceTypeId = @p0
        AND SearchParamId = @p1
        AND Code = @p2
),
cte1 AS
(
    SELECT DISTINCT TOP (@p3) Sid1, 1 AS IsMatch, 0 AS IsPartial 
    FROM cte0
    ORDER BY  Sid1 ASC
)
SELECT r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(IsMatch AS bit) AS IsMatch, CAST(IsPartial AS bit) AS IsPartial, CAST(0 AS bit) AS IsRawResourceMetaSet, r.RawResource
FROM dbo.Resource r
INNER JOIN cte1
ON r.ResourceSurrogateId = cte1.Sid1
ORDER BY r.ResourceSurrogateId ASC
OPTION(RECOMPILE)',N'@p0 smallint,@p1 smallint,@p2 varchar(128),@p3 int',@p0=81,@p1=861,**@p2='186cb836-2abe-4d3b-a928-325e2,5c8ca1c'**,@p3=11

as expected.

**Comment by johnstairs** [comment]
If you write your search query like this:
```
/Patient?identifier=XXXX\|YYYYYYY\|ZZZZZ
```
You will find the resource
